### PR TITLE
optimistic state simulating

### PR DESCRIPTION
**Pull request by chrisfarms**
## what

Introduces a way to quickly apply pending action transactions to a simulated fork of the target chain and update a fork of the indexed state so that the effects of transactions can we queryed before they are actually commited to main chain.

## resolves

* graph event subscriptions have been removed and replaced with a single block event (fixes: #16)
* dry-run has been removed, and replaced with a mechanism to fork and replay the pending/processing tx during dispatch (fixes: #11)
* revert reasons are now decoded and returned (fixes: #13)

## api changes

You can now request state at any block

```graphql
{
  games {    
    state(block: 2) {
      ...
    }
  }
}
```

You can request if you want to read the "real" state, or the latest optimisticly applied state

```graphql
{
  games {    
    state(simulated: true) {
      ...
    }
  }
}
```

## sequencer changes

* target chain will apply pending changes roughly every 10s by default
* dispatches will immediately apply changes to the simulated state

## contract changes

Converted use of "custom error" type to reason strings to avoid needing abi to decode 😢

### CODE REVIEW

File: services/pkg/client/alchemy/alchemy_client.go [hunk 1]
```diff
@@ -123,20 +123,38 @@ func (c *Client) NewRelayTransactor(ctx context.Context) (*bind.TransactOpts, er
...
 		"jsonRpcUrl":  remoteURL,
 		"blockNumber": remoteBlockNumber,
 	}
 	params := map[string]interface{}{
 		"forking": forking,
 	}
-	var hex hexutil.Uint64
-	err := c.rpc.CallContext(ctx, &hex, "anvil_reset", params)
+	var res json.RawMessage
+	err := c.rpc.CallContext(ctx, &res, "anvil_reset", params)
```
Comment (5p0rt5BEArD) [inline services/pkg/client/alchemy/alchemy_client.go:135]: Presumably making calls like anvil_reset and anvil_mine are specific to Foundry tools and these functions will need OP Stack (geth?) versions when we move?

File: services/pkg/sequencer/sequencer.go [hunk 2]
```diff
@@ -61,21 +64,26 @@ type MemorySequencer struct {
...
+	simWsClient   *alchemy.Client
+	simBlock      uint64
 	notifications chan interface{}
+	idxr          indexer.Indexer
 	log           zerolog.Logger
 	sync.RWMutex
 }
 
-func NewMemorySequencer(ctx context.Context, key *ecdsa.PrivateKey, notifications chan interface{}) (*MemorySequencer, error) {
+func NewMemorySequencer(ctx context.Context, key *ecdsa.PrivateKey, notifications chan interface{}, httpProviderURL string, simProviderURL string, idxr indexer.Indexer) (*MemorySequencer, error) {
```
Comment (5p0rt5BEArD) [inline services/pkg/sequencer/sequencer.go:76]: I'm trying to find where the simulation chain is actually connected to . I thought this new simProviderURL param would be the path to follow but its unused?

**Review by 5p0rt5BEArD — COMMENTED** [review]
I've been browsing changes to try and understand. Not really a review. I've asked a couple of questions that might help my understanding.

File: services/pkg/client/alchemy/alchemy_client.go (hunk 1 above)
Comment (chrisfarms) [reply services/pkg/client/alchemy/alchemy_client.go:135]: yes they are specific to anvil ... but anvil would always be used as the simulation fork here, regardless of upstream chain ... so they would/could stay as is

File: services/pkg/sequencer/sequencer.go (hunk 2 above)
Comment (chrisfarms) [reply services/pkg/sequencer/sequencer.go:76]: good spot ... looks like I was midway through trying to remove the use of the global `config` thing ... willfix

Comment (chrisfarms) [reply services/pkg/sequencer/sequencer.go]: fixed ... correctly uses the param now

File: services/pkg/config/config.go [hunk 3]
```diff
@@ -12,4 +12,7 @@ var SequencerPrivateKey = getRequiredEnvKey("SEQUENCER_PRIVATE_KEY")
 var SequencerMaxConcurrency = getOptionalEnvInt("SEQUENCER_MAX_CONCURRENCY", 200)
 var SequencerMinBatchDelayMilliseconds = getOptionalEnvInt("SEQUENCER_MIN_BATCH_DELAY_MS", 100)
 
+var SimulationProviderHTTP = getRequiredEnvString("SIMULATION_PROVIDER_URL_HTTP")
```
Comment (5p0rt5BEArD) [inline services/pkg/config/config.go:15]: Where are these new env strings set?

File: services/pkg/sequencer/sequencer.go (hunk 2 above)
Comment (5p0rt5BEArD) [reply services/pkg/sequencer/sequencer.go]: Thanks I've followed the code now through to ResetSim to where the sim urls are used.

**Comment by 5p0rt5BEArD** [comment]
Are you using the option to retrieve state as simulated or "real" in ds?

File: services/pkg/config/config.go (hunk 3 above)
Comment (chrisfarms) [reply services/pkg/config/config.go:15]: they would be set wherever you are spinning up cog services ... so in for ds that would be either:

* the .devstartup.js script that helps you with a local deployment
* the helm chart that configures a kubernetes deployment
* manually on the command line if you are manually running things

**Comment by chrisfarms** [comment]
> Are you using the option to retrieve state as simulated or "real" in ds?

ds will be always watching the simulated state .... but this change will enable us to replace the fake block number guessing used in combat with a subscription to the real block events as well as watching the simulated state for prettymuch everything else

**Review by 5p0rt5BEArD — APPROVED** [review]

**Pull request merged**